Sources/FoundationEssentials/AttributedString/DiscontiguousAttributedSubstring.swift

```diff
@@ -0,0 +1,320 @@
+extension DiscontiguousAttributedSubstring : Equatable {
+    public static func == (lhs: Self, rhs: Self) -> Bool {
+        AttributedString.Guts._characterwiseIsEqual(lhs.runs, to: rhs.runs)
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension DiscontiguousAttributedSubstring : Hashable {
+    public func hash(into hasher: inout Hasher) {
+        for range in _indices.ranges {
+            _guts.characterwiseHash(in: range, into: &hasher)
+        }
+    }
+}
+
+@available(FoundationPreview 6.2, *)
+extension DiscontiguousAttributedSubstring : AttributedStringAttributeMutation {
+    internal mutating func ensureUniqueReference() {
+        // Note: slices should never discard the data outside their bounds, so we must make a
+        // copy of the entire base string here.
+        //
+        // (Discarding out-of-range data would change index values, interfere with "in-place"
+        // mutations of slices via the subscript accessors, and it would confuse the semantics of
+        // the `base` property.)
+        if !isKnownUniquelyReferenced(&_guts) {
+            _guts = _guts.copy()
+        }
+    }
+    
+    public mutating func setAttributes(_ attributes: AttributeContainer) {
+        ensureUniqueReference()
+        for range in _indices.ranges {
+            _guts.setAttributes(attributes.storage, in: range)
+        }
+    }
+    
+    public mutating func mergeAttributes(_ attributes: AttributeContainer, mergePolicy:  AttributedString.AttributeMergePolicy = .keepNew) {
+        ensureUniqueReference()
+        for range in _indices.ranges {
+            _guts.mergeAttributes(attributes, in: range, mergePolicy: mergePolicy)
+        }
+    }
+    
+    public mutating func replaceAttributes(_ attributes: AttributeContainer, with others: AttributeContainer) {
+        guard attributes != others else {
+            return
+        }
+        ensureUniqueReference()
+        let hasConstrainedAttributes = attributes.storage.hasConstrainedAttributes || others.storage.hasConstrainedAttributes
+        var fixupRanges = [Range<Int>]()
+        for range in _indices.ranges {
+            _guts.runs(in: range._utf8OffsetRange).updateEach(
+                when: { $0.matches(attributes.storage) },
+                with: { runAttributes, utf8Range in
+                    for key in attributes.storage.keys {
+                        runAttributes[key] = nil
+                    }
+                    runAttributes.mergeIn(others)
+                    if hasConstrainedAttributes {
+                        fixupRanges.append(utf8Range)
+                    }
+                })
+        }
+        for range in fixupRanges {
+            // FIXME: Collect boundary constraints.
+            _guts.enforceAttributeConstraintsAfterMutation(in: range, type: .attributes)
+        }
+    }
+    
+    public subscript(bounds: some RangeExpression<AttributedString.Index>) -> DiscontiguousAttributedSubstring {
```

Comment:
Is this a new API or a protocol requirement? It doesn't look like a protocol requirement for `AttributedStringAttributeMutation`. Am I missing something?